### ISSUE TITLE
tf.keras batch normalization is batch dependent at test time

### ISSUE THREAD

**Issue by isaacgerg** [tensorflow/tensorflow#32544]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Both
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 7
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.13
- Python version: 3.7
- CUDA/cuDNN version: 10
- GPU model and memory: RTX

**Describe the current behavior**

When doing inference using tf.keras batchnormalization,  for a given input sample x, the output is dependent on the other samples in the batch.  I can modify the other samples in the batch to influence the output of x.  This should not be the case as all samples should be processed independently.  Whether I run a sample x as a batch size of 1 or with other samples, its output should be exactly the same.  I am seeing differences of 0.3 at times (see my output below).

It appears keras is also having this issue: https://github.com/keras-team/keras/issues/12400

**Describe the expected behavior**
Upon using tf.keras batch normalization at test time, a given input sample x should have the same output regardless of the other samples in the batch

**Code to reproduce the issue**
Note: I put the code into colab and was able to verify the bug.

```
import tensorflow as tf
import numpy as np

input1 = tf.keras.layers.Input(shape=(128,128,1))
x = tf.keras.layers.BatchNormalization()(input1)
x = tf.keras.layers.Flatten()(x)
x = tf.keras.layers.Dense(5, activation = 'softmax')(x)

model = tf.keras.models.Model(input1, x)

batchSize = 64
x = np.random.rand(1, 128,128,1)
x = np.repeat(x, batchSize, axis=0)

print('The following rows should all be equal...')
for k in range(1, batchSize):  
    y = model.predict(x[0:k,:,:,:], batch_size=8)
    print(y[0,:])
```
I get the following output:

```
The following rows should all be equal...
2019-09-15 16:46:29.463998: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1512] Adding visible gpu devices: 0
2019-09-15 16:46:29.463998: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-09-15 16:46:29.463998: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990]      0 
2019-09-15 16:46:29.463998: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1003] 0:   N 
2019-09-15 16:46:29.464999: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 11334 MB memory) -> physical GPU (device: 0, name: TITAN X (Pascal), pci bus id: 0000:09:00.0, compute capability: 6.1)
2019-09-15 16:46:30.597112: I tensorflow/stream_executor/dso_loader.cc:152] successfully opened CUDA library cublas64_100.dll locally
[0.38215435 0.66578805 0.45068434 0.85698235 0.7698223 ]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215438 0.66578805 0.45068428 0.85698223 0.76982236]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
[0.38215443 0.66578794 0.4506843  0.85698223 0.76982224]
```
When i run on the CPU, the problem goes away.

When I run on my own data, I get the following code / output which shows the problem more severely:
    
```
    x,y = testGenerator.next(0)
    score = []
    for kk in range(1, 16):
        s = model.predict(x[0:kk,:,:,:])
        print(s[0][0,:])
```
Gives output...
```
[9.3195647e-01 4.5844557e-04 6.6464258e-05 6.7509413e-02 2.0588086e-06
 1.3121526e-06 1.7270798e-06 2.0374323e-06 2.1472629e-06]
[9.3195647e-01 4.5844595e-04 6.6464192e-05 6.7509346e-02 2.0587986e-06
 1.3121462e-06 1.7270781e-06 2.0374266e-06 2.1472567e-06]
[9.3195623e-01 4.5845023e-04 6.6464687e-05 6.7509525e-02 2.0588097e-06
 1.3121534e-06 1.7270876e-06 2.0374339e-06 2.1472645e-06]
[9.1518945e-01 5.7900354e-04 8.6429020e-05 8.4134214e-02 2.4391386e-06
 1.5495950e-06 2.0541604e-06 2.3649241e-06 2.5034813e-06]
[9.1518945e-01 5.7900301e-04 8.6429180e-05 8.4134296e-02 2.4391431e-06
 1.5496009e-06 2.0541527e-06 2.3649197e-06 2.5034813e-06]
[9.1442782e-01 5.8447709e-04 8.7354209e-05 8.4889315e-02 2.4558769e-06
 1.5600473e-06 2.0685882e-06 2.3791872e-06 2.5190409e-06]
[9.1442782e-01 5.8447709e-04 8.7354209e-05 8.4889315e-02 2.4558769e-06
 1.5600473e-06 2.0685882e-06 2.3791872e-06 2.5190409e-06]
[9.1442782e-01 5.8447709e-04 8.7354209e-05 8.4889315e-02 2.4558769e-06
 1.5600473e-06 2.0685882e-06 2.3791872e-06 2.5190409e-06]
[9.1442782e-01 5.8447709e-04 8.7354209e-05 8.4889315e-02 2.4558769e-06
 1.5600473e-06 2.0685882e-06 2.3791872e-06 2.5190409e-06]
[2.7099210e-01 5.2615767e-03 1.4831917e-03 7.2222257e-01 9.9957060e-06
 6.1737474e-06 8.7498120e-06 7.3227852e-06 8.3933583e-06]
[2.7099210e-01 5.2615767e-03 1.4831917e-03 7.2222257e-01 9.9957060e-06
 6.1737474e-06 8.7498120e-06 7.3227852e-06 8.3933583e-06]
[2.7099210e-01 5.2615767e-03 1.4831917e-03 7.2222257e-01 9.9957060e-06
 6.1737474e-06 8.7498120e-06 7.3227852e-06 8.3933583e-06]
[2.7099210e-01 5.2615767e-03 1.4831917e-03 7.2222257e-01 9.9957060e-06
 6.1737474e-06 8.7498120e-06 7.3227852e-06 8.3933583e-06]
[2.7099210e-01 5.2615767e-03 1.4831917e-03 7.2222257e-01 9.9957060e-06
 6.1737474e-06 8.7498120e-06 7.3227852e-06 8.3933583e-06]
[1.9735371e-01 5.5795610e-03 1.7620743e-03 7.9526496e-01 9.8536966e-06
 6.0714729e-06 8.6973323e-06 7.0008568e-06 8.0944292e-06]
```

**EDIT 1:** Added code to replicate issue.  Also, If you remove the batch norm, the problem still persists. 

**EDIT 2:** Displayed results from running on my own data.

**Comment by gadagashwini** [comment]
@isaacgerg, I could execute the code in colab. Please take a look at colab [gist](https://colab.sandbox.google.com/gist/gadagashwini/ccbd9e29de1c4dd5cf591c08b18bbe41/untitled151.ipynb) and confirm the expected behavior. Thanks!

**Comment by isaacgerg** [comment]
@gadagashwini Yes, that gist looks correct.  For the run I did, the first 2 rows were not identical.

**Comment by twinanda** [comment]
Coming from https://github.com/keras-team/keras/issues/12400
I can verify that the rows are not equal for me. I tried multiple times, it is not equal.

**Comment by isaacgerg** [comment]
@gadagashwini I am able to replicate the error exactly when fixing the RNG seeds.

```
import random
import tensorflow as tf
import numpy as np

tf.keras.backend.clear_session()
random.seed(1)
np.random.seed(1)
tf.random.set_random_seed(1) 
input1 = tf.keras.layers.Input(shape=(128,128,1))
#x = tf.keras.layers.BatchNormalization()(input1)
x = tf.keras.layers.Flatten()(input1)
x = tf.keras.layers.Dense(5, activation = 'softmax')(x)

model = tf.keras.models.Model(input1, x)

batchSize = 32
x = np.random.rand(1, 128,128,1)
x = np.repeat(x, batchSize, axis=0)

print('The following rows should all be equal...')
for k in range(1, batchSize):  
    y = model.predict(x[0:k,:,:,:], batch_size=8)
    print(y[0,:])
```

**Comment by gadagashwini** [comment]
@isaacgerg, Can you please share the colab gist here. Thanks!

**Comment by isaacgerg** [comment]
@gadagashwini 
https://colab.research.google.com/gist/isaacgerg/5e7cd6edf27f8a7fe877a2628bc5fe65/tfkeras_bug1.ipynb

**Comment by isaacgerg** [comment]
@twinanda have you come up with a fix?

**Comment by twinanda** [comment]
@isaacgerg no, unfortunately not. I am looking into other normalization methods for this. Maybe instancenorm might yield similar convergence, without the inconsistency of batchnorm. Let me know if you find something on your side. :)

**Comment by OverLordGoldDragon** [comment]
I believe I've figured it out, but noone, myself included, will like the verdict: the problem is **numeric precision**, and `BatchNormalization` has nothing to do with it. It's a work in progress, but I've narrowed down the exact operation responsible: `np.dot` - you can follow the work [here](https://stackoverflow.com/questions/58740925/why-is-np-dot-imprecise-n-dim-arrays). Will update later with all relevant testing code, and a solution.

**Comment by isaacgerg** [comment]
@OverLordGoldDragon Thanks for following up.  I wonder if this is the same issue found with softmax here: https://github.com/keras-team/keras/issues/13328

**Comment by robieta** [comment]
I'm inclined to agree that this looks like normal floating point variability. Thanks for the analysis everyone.
